action/protocol/staking/handlers_test.go

```diff
@@ -0,0 +1,217 @@
+		// expected result
+		errorCause error
+	}{
+		{
+			0,
+			candidateName,
+			"10",
+			1,
+			false,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			state.ErrNotEnoughBalance,
+		},
+		{
+			100,
+			"notExist",
+			"10",
+			1,
+			false,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			ErrInvalidCanName,
+		},
+		{
+			100,
+			candidateName,
+			"10000000000000000000",
+			1,
+			false,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			nil,
+		},
+	}
+
+	for _, test := range tests {
+		require.NoError(setupAccount(sm, stakerAddr, test.initBalance))
+		ctx := protocol.WithActionCtx(context.Background(), protocol.ActionCtx{
+			Caller:       stakerAddr,
+			GasPrice:     test.gasPrice,
+			IntrinsicGas: test.gasLimit,
+			Nonce:        test.nonce,
+		})
+		ctx = protocol.WithBlockCtx(ctx, protocol.BlockCtx{
+			BlockHeight:    test.blkHeight,
+			BlockTimeStamp: test.blkTimestamp,
+			GasLimit:       test.blkGasLimit,
+		})
+		act, err := action.NewCreateStake(test.nonce, test.candName, test.amount, test.duration, test.autoStake,
+			nil, test.gasLimit, test.gasPrice)
+		require.NoError(err)
+		_, err = p.handleCreateStake(ctx, act, sm)
+		require.Equal(test.errorCause, errors.Cause(err))
+
+		if test.errorCause == nil {
+			// test bucket index and bucket
+			bucketIndices, err := getCandBucketIndices(sm, candidateAddr)
+			require.NoError(err)
+			require.Equal(1, len(*bucketIndices))
+			bucketIndices, err = getVoterBucketIndices(sm, stakerAddr)
+			require.NoError(err)
+			require.Equal(1, len(*bucketIndices))
+			indices := *bucketIndices
+			bucket, err := getBucket(sm, indices[0])
+			require.NoError(err)
+			require.Equal(candidateAddr, bucket.Candidate)
+			require.Equal(stakerAddr, bucket.Owner)
+			require.Equal(test.amount, bucket.StakedAmount.String())
+
+			// test candidate
+			candidate, err := getCandidate(sm, candidateAddr)
+			require.NoError(err)
+			require.LessOrEqual(test.amount, candidate.Votes.String())
+			candidate = p.inMemCandidates.GetByOwner(candidateAddr)
+			require.NotNil(candidate)
+			require.LessOrEqual(test.amount, candidate.Votes.String())
+
+			// test staker's account
+			caller, err := accountutil.LoadAccount(sm, hash.BytesToHash160(stakerAddr.Bytes()))
+			require.NoError(err)
+			actCost, err := act.Cost()
+			require.NoError(err)
+			require.Equal(unit.ConvertIotxToRau(test.initBalance), big.NewInt(0).Add(caller.Balance, actCost))
+			require.Equal(test.nonce, caller.Nonce)
+		}
+	}
+}
+
+func setupAccount(sm protocol.StateManager, addr address.Address, balance int64) error {
+	if balance < 0 {
+		return errors.New("balance cannot be negative")
+	}
+	account, err := accountutil.LoadOrCreateAccount(sm, addr.String())
+	if err != nil {
+		return err
+	}
+	account.Balance = unit.ConvertIotxToRau(balance)
+	return accountutil.StoreAccount(sm, addr.String(), account)
+}
+
+func setupCandidate(
```

Comment:
change to setup(p, sm, *Candidate)

and use one of testCandidates (already defined)